extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -0,0 +1,617 @@
+    props.setProperty("metadata.max.age.ms", "10000");
+    props.setProperty("bootstrap.servers", bootstrapServers);
+    props.setProperty("group.id", String.format("kafka-supervisor-%s", supervisorRunId));
+
+    ClassLoader currCtxCl = Thread.currentThread().getContextClassLoader();
+    try {
+      Thread.currentThread().setContextClassLoader(getClass().getClassLoader());
+      return new KafkaConsumer<>(props, new ByteArrayDeserializer(), new ByteArrayDeserializer());
+    }
+    finally {
+      Thread.currentThread().setContextClassLoader(currCtxCl);
+    }
+  }
+
+  private void updatePartitionDataFromKafka()
+  {
+    try {
+      consumer.listTopics(); // updates the consumer's list of partitions from the brokers
+    }
+    catch (Exception e) { // calls to the consumer throw NPEs when the broker doesn't respond
+      log.warn(
+          e,
+          "Unable to get partition data from Kafka for brokers [%s], are the brokers up?",
+          spec.getIoConfig().getKafkaBrokers()
+      );
+      return;
+    }
+
+    List<PartitionInfo> partitions = consumer.partitionsFor(spec.getIoConfig().getTopic());
+    int numPartitions = (partitions != null ? partitions.size() : 0);
+
+    log.debug("Found [%d] Kafka partitions for topic [%s].", numPartitions, spec.getIoConfig().getTopic());
+
+    for (int partition = 0; partition < numPartitions; partition++) {
+      int taskGroupId = getTaskGroupIdForPartition(partition);
+      if (taskGroups.get(taskGroupId) == null) {
+        log.info("Creating new task group [%d]", taskGroupId);
+        taskGroups.put(taskGroupId, new TaskGroup());
+      }
+
+      if (!taskGroups.get(taskGroupId).partitions.containsKey(partition)) {
+        log.info(
+            "New partition [%d] discovered for topic [%s], adding to task group [%d]",
+            partition,
+            spec.getIoConfig().getTopic(),
+            taskGroupId
+        );
+        taskGroups.get(taskGroupId).partitions.put(partition, new PartitionOffsets(0, 0, false));
+      }
+    }
+  }
+
+  private void updateTaskMetadata()
+  {
+    int taskCount = 0;
+    List<Task> tasks = taskStorage.getActiveTasks();
+
+    for (Task task : tasks) {
+      if (task instanceof KafkaIndexTask && dataSource.equals(task.getDataSource())) {
+        taskCount++;
+        KafkaIndexTask kafkaTask = (KafkaIndexTask) task;
+
+        // determine which task group this task belongs to and do a consistency check on partitions
+        Integer taskGroupId = null;
+        for (Integer partition : kafkaTask.getIOConfig().getStartPartitions().getPartitionOffsetMap().keySet()) {
+          if (taskGroupId == null) {
+            taskGroupId = getTaskGroupIdForPartition(partition);
+          } else if (taskGroupId != getTaskGroupIdForPartition(partition)) {
+            log.error("Discovered task [%s] has different partition allocation than expected!", task.getId());
+            taskGroupId = null;
+            break;
```

Comment:
If we're really expecting this case to never happen, perhaps instead of breaking out of this one task, we should instead emit an alert and throw an exception (aborting the entire management run)